### DISCUSSION TITLE
2022.622.0

### DISCUSSION THREAD

**Discussion by VolcanicArts** [VolcanicArts/VRCOSC#10] [Releases]
# Installation
- Download VRCOSCSetup.exe (It may get blocked by Windows Defender)
- Run VRCOSCSetup.exe. A shortcut will be made on your desktop and in your start menu
- When adding any of the OSC prefabs to your avatar, after uploading you need to reset the OSC config from your action menu
- If you encounter a problem of any sort with this release, check out the discussion created in the [Discussions](https://github.com/VolcanicArts/VRCOSC/discussions/categories/releases) tab

# What's Changed
- `OSCClient` logic improvements
- Added option to stop Spotify on `SpotifyModule` stop
- Refactored `IntegrationModule` to be more efficient and safe
- Allow `IntegrationModule` to ensure only a single process exists before executing code
- Allow toggling of global setting `AutoStartStop` to re-enable auto start/stop during runtime if the user has manually stopped running the modules
- Attempted fix for the unresponsive first button push on all integration modules
  - Properly awaiting the task delays should help on slower systems, but it still seems like Windows doesn't focus a process immediately when accessing it from code
- Created `TimedTask` to deal with timed tasks better
- Converted `HypeRateProvider` to use `TimedTask`
- Converted `Module` to use `TimedTask` for the `OnUpdate` event
- Fixed a regression of `DeltaUpdate` which caused all `OnUpdate` events to be called even if `DeltaUpdate` was max value
  - This was meant to be set to `double.PositiveInfinity` to properly cancel calling `OnUpdate` but it was set to `double.MaxValue` during the `Module` refactor
- Marked `DeltaUpdate` as `protected` instead of `public` to communicate that it is read at runtime and not `Module` instantiation

**Full Changelog**: https://github.com/VolcanicArts/VRCOSC/compare/2022.621.0...2022.622.0

<hr /><em>This discussion was created from the release <a href='https://github.com/VolcanicArts/VRCOSC/releases/tag/2022.622.0'>2022.622.0</a>.</em>

**Comment by Lucy231k** [comment]
Hi is there a video tutorial or a discord to join and get help from?

**Reply by VolcanicArts** [reply]
There isn't anything like that, no. But each prefab contains a guide on how to put them on your avatar if you're interested in them. I can give you a quick rundown here though.

If you download "VRCOSCSetup.exe" and run it, it will install. A shortcut will be placed on your desktop so you can easily access it again.

If, for example, you wanted the watch prefab, download "VRCOSC-Watch-2022.609.0.unitypackage". The "Guide.txt" will help you install it onto your avatar, remembering that it needs to be placed on the root for the animations to work correctly (This is the case for all prefabs that aren't just for the action menu).

After uploading your avatar, go into your action menu under "options", and "osc", and you'll see a "reset config" button. Press that.

Start VRCOSC and click the red power icon on the "Clock" module, this will turn green and enable it.
Clicking the cog icon will take you to the "Clock" module's settings and parameters. You only need to worry about the options under the "Settings" title for customisation.

Then you can either press the run button in the bottom left, or enable the "Auto Start/Stop" setting in the top left, which will auto start VRCOSC's enabled modules if VRC is running.

If you have anymore questions feel free to add me on Discord: `VolcanicArts#0001`.
I'm more than happy to run you through things in more details. I have been meaning to make the guides into visual PDFs.
